pkg/rkt-stage1/0011-Patch-to-mount-vol-stage1.patch

```diff
@@ -0,0 +1,302 @@
++	if err != nil {
++		return nil, errwrap.Wrap(errors.New("can't gather host mounts"), err)
++	}
++	for _, m := range mnts {
++		hostMounts[m.MountPoint] = struct{}{}
++	}
++
++	argXenMounts := []xenMount{}
++	for _, tuple := range namedVolumeMounts {
++		if _, isHostMount := hostMounts[tuple.V.Source]; isHostMount {
++			// Mark the host mount as SHARED so the container's changes to the mount are propagated to the host
++			argXenMounts = append(argXenMounts,
++				xenMount{"", "", tuple.V.Source, "none", syscall.MS_REC | syscall.MS_SHARED},
++			)
++		}
++
++		var (
++			flags     uintptr = syscall.MS_BIND
++			recursive         = tuple.V.Recursive != nil && *tuple.V.Recursive
++			ro                = tuple.V.ReadOnly != nil && *tuple.V.ReadOnly
++		)
++
++		// If Recursive is not set, default to non-recursive.
++		if recursive {
++			flags |= syscall.MS_REC
++		}
++
++		argXenMounts = append(argXenMounts,
++			xenMount{tuple.V.Source, rfs, tuple.M.Path, "none", flags},
++		)
++
++		if ro {
++			argXenMounts = append(argXenMounts,
++				xenMount{"", rfs, tuple.M.Path, "none", flags | syscall.MS_REMOUNT | syscall.MS_RDONLY},
++			)
++
++			if recursive {
++				// Every sub-mount needs to be remounted read-only separately
++				mnts, err := mountinfo.ParseMounts(0)
++				if err != nil {
++					return nil, errwrap.Wrap(fmt.Errorf("error getting mounts under %q from mountinfo", tuple.V.Source), err)
++				}
++				mnts = mnts.Filter(mountinfo.HasPrefix(tuple.V.Source + "/"))
++
++				for _, mnt := range mnts {
++					innerRelPath := tuple.M.Path + strings.Replace(mnt.MountPoint, tuple.V.Source, "", -1)
++					argXenMounts = append(argXenMounts,
++						xenMount{"", rfs, innerRelPath, "none", flags | syscall.MS_REMOUNT | syscall.MS_RDONLY},
++					)
++				}
++			}
++		}
++	}
++	return argXenMounts, nil
++}
++
+ // getArgsEnv returns the nspawn or lkvm args and env according to the flavor
+ // as the first two return values respectively.
+ func getArgsEnv(p *stage1commontypes.Pod, flavor string, debug bool, n *networking.Networking) ([]string, []string, error) {
+@@ -204,6 +358,109 @@ func stage1(rp *stage1commontypes.RuntimePod) int {
+ 	}
+
+ 	ra := p.Manifest.Apps[0]
++	rfs := filepath.Join(common.AppPath(p.Root, ra.Name), "rootfs")
++
++	argFlyMounts, err := evaluateMounts(rfs, string(ra.Name), p)
++	if err != nil {
++		log.PrintE("can't evaluate mounts", err)
++		return 254
++	}
++
++	effectiveMounts := append(
++		[]xenMount{
++			{"", "", "/dev", "none", syscall.MS_REC | syscall.MS_SHARED},
++			{"/dev", rfs, "/dev", "none", syscall.MS_BIND | syscall.MS_REC},
++
++			{"", "", "/proc", "none", syscall.MS_REC | syscall.MS_SHARED},
++			{"/proc", rfs, "/proc", "none", syscall.MS_BIND | syscall.MS_REC},
++
++			{"", "", "/sys", "none", syscall.MS_REC | syscall.MS_SHARED},
++			{"/sys", rfs, "/sys", "none", syscall.MS_BIND | syscall.MS_REC},
++
++			{"tmpfs", rfs, "/tmp", "tmpfs", 0},
++		},
++		argFlyMounts...,
++	)
```

Comment:
This still appears to be more complex than it needs to be: why are we mimicking container-specific part of this? Its not like Xen will ever see any of this, right?